lib/ueberauth/strategy/facebook.ex

```diff
@@ -149,13 +149,26 @@ defmodule Ueberauth.Strategy.Facebook do
     end
   end
 
-  defp user_query(conn) do
-    conn
-    |> query_params(:locale)
+  defp user_query(conn, token) do
+    %{"appsecret_proof_" => appsecret_proof(token)}
+    |> Map.merge(query_params(conn, :locale))
     |> Map.merge(query_params(conn, :profile))
     |> URI.encode_query
   end
 
+  defp appsecret_proof(token) do
+    config = Application.get_env(:ueberauth, Ueberauth.Strategy.Facebook.OAuth)
+    client_secret = Keyword.get(config, :client_secret)
+
+    token.access_token
+    |> hmac(:sha256, client_secret)
+    |> Base.encode16(case: :lower)
+  end
+
+  defp hmac(data, type, key) do
```

Comment:
If you wonder why I defined a helper for this it's to improve the pipeline readability: `data |> hmac(type, key)` reads better (like `data |> Base.encode16`) than `type |> :crypto.hmac(key, data)`. Elixir should really wrap those Erlang natives.